### PR TITLE
Add server logging interface

### PR DESCRIPTION
**Pull request by mrogers950**
@simo5 @enj We should do some nitpicking over the exact data to log in each case to make sure there's nothing we should not be logging and that the context given to each makes sense.

I also fixed util_test.go and access_test.go compilation (I can break this to another commit/pr if needed)

### CODE REVIEW

File: access.go [hunk 1]
```diff
@@ -525,38 +511,58 @@ func (s *Server) FinishAccessRequest(w *Response, r *http.Request, ar *AccessReq
...
-		w.SetError(E_UNAUTHORIZED_CLIENT, "")
+		s.setErrorAndLog(w, E_UNAUTHORIZED_CLIENT, "client", "no redirect uri")
 		return nil
 	}
 	return client
 }
+
+// setErrorAndLog sets the response error, and calls the Logger with additional debug data.
+func (s Server) setErrorAndLog(w *Response, responseError string, debug ...string) {
+	w.SetError(responseError, "")
```
Comment (simo5) [inline access.go]: In some instances the original code was passing an error string here, now it never happens, why is this ok ?

File: access.go [hunk 2]
```diff
@@ -115,20 +114,17 @@ func (s *Server) HandleAccessRequest(w *Response, r *http.Request) *AccessReques
...
-		w.InternalError = errors.New("Request must be POST")
+		s.setErrorAndLog(w, E_INVALID_REQUEST, "request_method_error", "request must be POST")
 		return nil
 	}
 
 	err := r.ParseForm()
 	if err != nil {
-		w.SetError(E_INVALID_REQUEST, "")
-		w.InternalError = err
+		s.setErrorAndLog(w, E_INVALID_REQUEST, "form_parsing_error", err.Error())
```
[the remark this answers is not in the document]
Comment (mrogers950) [reply access.go]: Ah, good point. I will check these out more closely.

**Comment by mrogers950** [comment]
@simo5 I've updated with changes that mirror the suggestions in the upstream PR (we have feedback from the maintainer now), but we may not ultimately need to merge this one.

WRT the SetError "description" argument @enj informed me that the three usages of description were actually unnecessary when adding the PKCE code, and we can use the default description for all calls of setErrorAndLog. Also, the internal errors should be OK to log in each call; OpenShift logs them regardless outside of osin.

**Comment by simo5** [comment]
/lgtm
Let's see if upstream takes it before we spend more time with this repo, if he does we may continue to use upstream.

**Comment by enj** [comment]
Merged upstream.

**Pull request closed without merging**